### ISSUE TITLE
combining an extra feature with glove embeddings

### ISSUE THREAD

**Issue by Ankit33-dev** [howardyclo/pytorch-seq2seq-example#4]
Hello,

First of all I would I like to thank you for this awesome explanation of seq2seq model. 
As I am very new to pytorch this detailed explanation is very much helpful.
I am trying to implement this model for text summarization task. The dataset I am using is Gigaword dataset. 
I trained it for 3 epochs and it gave me pretty good results. Next thing I want to do is to add an extra feature.
For every line in dataset I have extracted a contextual layer, which consists of subject, object and predicate for the sentence. I want to feed this along with the glove embeddings as a feature. How can I convert this contextual layer in a form which can be accepted by the model and is it wise to concatenate with the  tgt_embedding as there will be no src_embedding for this feature.

Looking forward to hear from you.
Thanks in advance.

**Comment by howardyclo** [comment]
@sunnycse Hi, I'm very glad to hear this repo helps you!
1. *"How can I convert this contextual layer in a form which can be accepted by the model?"*

If for me to implement this, in the encoding phase, I would add additional "feature embeddings" to embed the feature with the word. For example, your input source tokens
`["I", "love", "NLP" ,"."]` with the corresponding feature tokens `["Subject", "Predicate", "Object", "None"]`, the glove embedding (`nn.Embedding` object) first encodes `["I", "love", "NLP" ,"."]` to glove vectors with shape `(time_steps=4,  batch_size=1, word_vec_size)`. Second, you need to use another feature embedding (`nn.Embedding` object) encodes `["Subject", "Predicate", "Object", "None"]` to feature vectors with shape `(time_steps=4,  batch_size=1, feat_vec_size)`. Now, you can concatenate this two vectors to the shape `(time_steps=4,  batch_size=1, word_vec_size + feat_vec_size)` or maybe further do projection with `nn.Linear` into `(time_steps=4,  batch_size=1, projected_size)` and feed it into the encoder.

2. *"Is it wise to concatenate with the tgt_embedding as there will be no src_embedding for this feature."*

Sorry I couldn't fully understand the question. I will answer it as possible as I can. In inference stage, I think you don't know the ground-truth features of target tokens, so how do you concatenate target feature embeddings with target glove embeddings? Unless you also make decoder predict target features (but in my experience, multi-tasking usually hurts performance).

**Comment by Ankit33-dev** [comment]
@howardyclo 
Hello,
I really appreciate for your reply and thank you for the detailed explanation. 
As far as I understood, I have used pretrained golve embeddings and passed them input tokens i.e word2idx . 
and the glove vectors from above, I have extracted and passed num_embeddings and embedding_dim to nn.Embedding object.
One thing I am not able to understood, as you explained above is to convert glove vectors into shape (time_steps=4, batch_size=1, word_vec_size). What is the benefit of this and how can achieve this?.

My second question is, is it possible to pass multiple nn.Embedding object into encoder and decoder?
Will it be something like this?
class EncoderRNN(nn.Module):
   def __init__(self, embedding=None, **num_embeddings1, num_embeddings2,embedding_dim1,embedding_dim2**, rnn_type='LSTM', hidden_size=128, num_layers=1, dropout=0.3, bidirectional=True):

self.num_layers = num_layers
        self.dropout = dropout
        self.bidirectional = bidirectional
        self.num_directions = 2 if bidirectional else 1 
        self.hidden_size = hidden_size // self.num_directions        
        self.embedding = embedding
        self.word_vec_size = self.embedding.embedding_dim        
        self.rnn_type = rnn_type
        **self.feature2_embedding = nn.Embedding(num_embeddings1, embedding_dim1)
        self.feature3_embedding = nn.Embedding(num_embeddings2, embedding_dim2)**

Looking forward to hear from you.
Thankyou

**Comment by howardyclo** [comment]
@sunnycse  

1. *"One thing I am not able to understood, as you explained above is to convert glove vectors into shape (time_steps=4, batch_size=1, word_vec_size). What is the benefit of this and how can achieve this?."*

I think I were just trying to explain in detail by commenting the shape information so that you can get the idea. Don't care too much about the shape I said (LOL). I were just saying concatenating word vector and feature vector along the last dimension.

2. *"My second question is, is it possible to pass multiple nn.Embedding object into encoder and decoder? Will it be something like this? ...."*

Yes, sure! You can do something like that. For cleaner code, I will use `nn.ModuleList` or `nn.ModuleDict` to make all the embedding objects together. They are similar to `nn.Sequential` which you put multiple layers in it. See [*OpenNMT*'s embeddings implementation](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/modules/embeddings.py), I think that is what you want exactly.

**Comment by Ankit33-dev** [comment]
@howardyclo  Thanks alot for your reply. This will be very helpful :)
